controllers/openstacknetattachment_controller.go

```diff
@@ -0,0 +1,675 @@
+		if condition != nil {
+			if condition.Type == nmstateshared.NodeNetworkConfigurationPolicyConditionAvailable {
+				cond.Message = fmt.Sprintf("OpenStackNetAttach %s configured targeted node(s)", networkConfigurationPolicy.Name)
+				cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.NetAttachConfigured)
+
+			} else if condition.Type == nmstateshared.NodeNetworkConfigurationPolicyConditionDegraded {
+				cond.Message = fmt.Sprintf("underlying NNCP error: %s", condition.Message)
+				// set condition reason from nncp that we can react on CR cleanup
+				cond.Reason = ospdirectorv1beta1.ConditionReason(condition.Reason)
+				cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.NetAttachError)
+
+				return common.WrapErrorForObject(cond.Message, instance, err)
+			}
+		}
+	}
+	common.LogForObject(r, cond.Message, instance)
+
+	return nil
+}
+
+func (r *OpenStackNetAttachmentReconciler) cleanupNodeNetworkConfigurationPolicy(
+	instance *ospdirectorv1beta1.OpenStackNetAttachment,
+	cond *ospdirectorv1beta1.Condition,
+) (ctrl.Result, error) {
+	networkConfigurationPolicy := &nmstatev1alpha1.NodeNetworkConfigurationPolicy{}
+
+	//
+	// Get the bridge name from CR desired state
+	//
+	desiredState, err := r.getNNCPDesiredState(
+		instance,
+		&instance.Spec.AttachConfiguration.NodeNetworkConfigurationPolicy,
+		cond,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	bridgeName := gjson.Get(string(desiredState["desiredState"]), "interfaces.#.name").Array()[0].String()
+
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: bridgeName}, networkConfigurationPolicy)
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	//
+	// Get the current desired state from nncp object
+	//
+	desiredState, err = r.getNNCPDesiredState(
+		instance,
+		&networkConfigurationPolicy.Spec,
+		cond,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	//
+	// Set/update CR status from NNCP status
+	//
+	err = r.getNodeNetworkConfigurationPolicyStatus(instance, cond)
+
+	// in case of nncp cond.Reason == FailedToConfigure, still continue to try to
+	// cleanup and delete the nncp
+	if err != nil &&
+		cond.Reason != ospdirectorv1beta1.ConditionReason(nmstateshared.NodeNetworkConfigurationPolicyConditionFailedToConfigure) {
+		return ctrl.Result{}, err
+	}
+
+	bridgeState := gjson.Get(string(desiredState["desiredState"]), "interfaces.#.state").Array()[0].String()
+
+	if bridgeState != "absent" && bridgeState != "down" {
+		apply := func() error {
+			//
+			// Update nncp desired state to down to unconfigure the device on the worker nodes
+			//
+			re := regexp.MustCompile(`"state":"up"`)
+			desiredStateAbsent := re.ReplaceAllString(string(desiredState["desiredState"]), `"state":"down"`)
+			networkConfigurationPolicy.Spec.DesiredState = nmstateshared.State{
+				Raw: nmstateshared.RawState(desiredStateAbsent),
+			}
```

Comment:
Ah, will this fix the problem we work around [here](https://github.com/openstack-k8s-operators/osp-director-dev-tools/blob/455f27cb7ad5b5a61cb742fc6d182d0c393c2b7b/ansible/openstack_cleanup.yaml#L23-L31) in dev-tools?

Reply:
yes, this is the proper way to remove a previously configured device. need to remove that linked part in my dev-tools patch